tools/analyzer/main.rs

```diff
@@ -0,0 +1,78 @@
+use std::env;
+use std::fs;
+use std::io::ErrorKind;
+use std::path::PathBuf;
+use std::process::Command;
+
+// TODO(david): This shells out to an expected rule in the workspace root //:rust_analyzer that the user must define.
+// It would be more convenient if it could automatically discover all the rust code in the workspace if this target does not exist.
+fn main() {
+    let repo_root = workspace_dir().expect(
+        "Could not determine workspace, are you running inside a directory with WORKSPACE?",
+    );
+    let bazel_command = env::var("BAZEL_COMMAND").unwrap_or_else(|_| "bazel".into());
+    let analyzer_target =
+        env::var("BAZEL_ANALYZER_TARGET").unwrap_or_else(|_| "//:rust_analyzer".into());
+
+    env::set_current_dir(&repo_root).expect(
+        format!(
+            "could not access workspace directory: {}",
+            repo_root.to_string_lossy()
+        )
+        .as_str(),
+    );
+    let output = Command::new(bazel_command)
+        .arg("build")
+        .arg(&analyzer_target)
+        .output()
+        .expect("failed to execute bazel process");
+    if !output.status.success() {
+        panic!(
+            "bazel build failed:({}) of {}:\n{}",
+            output.status,
+            analyzer_target,
+            std::str::from_utf8(&output.stderr).unwrap()
+        );
+    }
+
+    let generated_rust_project = repo_root.join("bazel-bin/rust-project.json");
+    let workspace_rust_project = repo_root.join("rust-project.json");
+    let canonical_exec_root = fs::canonicalize(repo_root.join(format!(
```

Comment:
As we're already invoking `bazel` in this program, can we run `bazel info execution_root` to get this path, rather than making assumptions about symlinks?

(If not, can you make an `OsString` (with `push`), rather than formatting into a `String`? Avoids needing the lossy conversion (I know we're going to do one later when we JSONify, but minimising their scope to just when we're crossing a serialization/deserialization boundary is nice :)))

Reply:
Done. I agree that `bazel info execution_root` is a better way to find this, but it's a bit unsatisfying because the command output is a `Vec<u8>` and there doesn't seem to be a cross-platform way to interpret that as an OsString (unix has an extension to do so, but windows doesn't seem to), so maybe it's better to go the `OsString` with `push` route? What do you think?